### PR TITLE
Block plan application until state store has caught up to raft

### PR DESCRIPTION
**Pull request by schmichael**
**Do not merge until 0.9.1**

This guards against a situation where planApply could dequeue a plan that references objects that are committed to raft but not yet applied to the leader's state store.

To do this I refactored `worker.waitForIndex` into `StateStore.SnapshotAfter` (new name welcome) so it could be shared between the 3 call sites. Backoff logic should be maintained although error log lines will vary slightly from the old waitForIndex errors ("timeout" vs "deadline exceeded").

## TODOs

* Unit test planApply - not sure how to control raft and state store's synchronization, but I haven't spent much time digging in. Advice welcome.
* Integration/performance test - This will be done in a follow up PR.
* Metrics and/or logging around the planApply callsite
    * Perhaps metrics should be internal to SnapshotAfter and we skip timing the fast path that doesn't block? That would create a nice metric that only measures any lag between the indices and not the non-blocking path.

### CODE REVIEW

File: nomad/worker.go [hunk 1]
```diff
@@ -224,56 +224,24 @@ func (w *Worker) sendAck(evalID, token string, ack bool) {
...
-	// Exponential back off if we haven't yet reached it
-	if w.backoffErr(backoffBaselineFast, backoffLimitFast) {
-		return fmt.Errorf("shutdown while waiting for state sync")
-	}
-	goto CHECK
+	ctx, cancel := context.WithTimeout(w.srv.shutdownCtx, timeout)
+	snap, err := w.srv.fsm.State().SnapshotAfter(ctx, waitIndex)
+	cancel()
+	metrics.MeasureSince([]string{"nomad", "worker", "wait_for_index"}, start)
+	return snap, err
```
Comment (dadgar) [inline nomad/worker.go]: Wrap this error, if it hits its timeout it will just be context cancelled. `fmt.Errorf("failed to snapshot state at index after %d: %v", waitIndex, err)`

File: nomad/state/notify.go [hunk 2]
```diff
@@ -1,62 +0,0 @@
-package state
```
Comment (preetapan) [inline nomad/state/notify.go]: why was this deleted

File: nomad/plan_apply.go [hunk 3]
```diff
@@ -99,7 +100,9 @@ func (p *planner) planApply() {
 		// Snapshot the state so that we have a consistent view of the world
 		// if no snapshot is available
 		if waitCh == nil || snap == nil {
-			snap, err = p.fsm.State().Snapshot()
+			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```
[the remark this answers is not in the document]
Comment (schmichael) [reply nomad/plan_apply.go]: Ah, that makes sense. Bumping to 5s for now.

File: nomad/state/notify.go (hunk 2 above)
Comment (schmichael) [reply nomad/state/notify.go]: it's completely unused

File: nomad/worker.go (hunk 1 above)
Comment (schmichael) [reply nomad/worker.go]: Good call. Fixed.
